src/torrent.cpp

```diff
@@ -5334,30 +5304,29 @@ namespace libtorrent
 	{
 		TORRENT_ASSERT(p != nullptr);
 		TORRENT_ASSERT(is_single_thread());
+		TORRENT_ASSERT(std::count(m_peers_to_disconnect.begin()
+			, m_peers_to_disconnect.end(), p) == 0);
 
-		auto const i = sorted_find(m_connections, p);
-		if (i == m_connections.end())
-		{
-			TORRENT_ASSERT_FAIL();
-			return;
-		}
+		std::weak_ptr<torrent> weak_t = shared_from_this();
+		m_peers_to_disconnect.push_back(p);
+		m_deferred_disconnect.post(m_ses.get_io_service(), [=]() {
```

Comment:
Now that I see the code, maybe `deferred_handler` could have some `should_post` and save the creation of the handler, that I suppose it's not free, or is there some magic in the `&&`?